doc/developers/js_app.rst

```diff
@@ -0,0 +1,167 @@
+JavaScript Application
+======================
+
+CCF includes a native ``js_generic`` application which can execute JavaScript applications proposed through governance.
+
+CCF's JavaScript environment is similar to that in browsers, with the following differences:
+
+- only `JavaScript modules <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules>`_ are supported,
+- no `Web APIs <https://developer.mozilla.org/en-US/docs/Web/API>`_ are available,
+- module imports cannot be URLs and must be relative paths.
+
+.. _js-app-bundle:
+
+App bundle
+----------
+
+The most basic CCF JavaScript app has the following folder structure:
+
+.. code-block:: bash
+
+    $ tree --dirsfirst my-app
+    my-app
+    ├── src
+    │   └── app.js
+    └── app.json
+
+We also call this a `JavaScript app bundle` since it is the format used for deployment.
+It consists of metadata (``app.json``) and one or more JavaScript modules (``src/``)
+-- more details in later sections.
+
+You can find an example app bundle in the
+`tests/js-app-bundle <https://github.com/microsoft/CCF/tree/master/tests/js-app-bundle>`_
+folder of the CCF git repository.
+
+An app bundle can be turned into a proposal with the Python client:
+
+.. code-block:: bash
+
+    $ python -m ccf.proposal_generator deploy_js_app my-app/
+    SUCCESS | Writing proposal to ./deploy_js_app_proposal.json
+    SUCCESS | Wrote vote to ./deploy_js_app_vote_for.json
+
+Once accepted, a ``deploy_js_app`` proposal atomically (re-)deploys the complete JavaScript application.
+Any existing application endpoints and JavaScript modules are removed.
+
+In the following sections we go into details of the metadata file and
```

Comment:
I think we should try to keep one sentence per line wherever possible?

Reply:
Hmm, even if the line becomes too long? I typically break them so that git diff works alright. Also, the VS Code diff view does not support word wrapping.

Reply:
My preference is one sentence per line as it's easier to read in VSCode when editing the file. I agree that it makes the diffs in VSCode a little impractical but GitHub renders the changes alright. No a big deal in any case, so feel free to leave it as it is.

Reply:
Let's leave it as it is for now. I'll probably make further changes to these docs anyway in the near future.